### ISSUE TITLE
Add `composer install --keep-removed` to help uninstalling modules in Drupal deployment

### ISSUE THREAD

**Issue by geek-merlin** [composer/composer#10824]
Coming from the Drupal world, a common challenge is the following:
To get rid of an obsoleted module component, we have to
- first `drush pm-uninstall mymodule` the module (which cleans up its data tables)
-  then `composer remove drupal/mymodule`

But if i do **both** on a dev site, and deploy to a live site like this:
```
composer install
drush config-import
```
i get the problem that
- first the module is removed by composer
- then the module's uninstall handler should be called, but is not there anymore

A common workaround is to do the second step only when all live sites are deployed with the first step. But this is cumbersome and a source of errors.

What would help a lot is to deploy like that:
```
# Adds and updates all packages, but does not delete removed ones.
composer install --keep-removed
drush config-import
# Actually deletes removed packages.
composer install
```
Maybe with a small code change this can help at least the Drupal ecosystem a lot, and maybe others with similar paradigms.

**Comment by Seldaek** [comment]
I guess the ideal way here would be if drupal/drush had a composer plugin that would deal with uninstalls/config-import automatically as the composer install happens, then it'd be one integrated process.

I realize this is a further-away goal than the workaround you ask for, but I am also not super keen on adding hacky flags like this that serve only a limited use case, and may induce others to do weird things instead of fixing things properly.

I'd also like to hear from some Drupal/Drush folks here before deciding on anything.

**Comment by Seldaek** [comment]
- Prevent deleting.. maybe. It'd require a bigger amount of hackery than your average plugin I guess, but it should be possible I'd think.
- Act on the list of changes that is to be done, definitely. There's `pre-operations-exec` or `pre-package-uninstall` from https://getcomposer.org/doc/articles/scripts.md which should allow a plugin to hook into this process.

**Comment by Seldaek** [comment]
Pinging @weitzman and @greg-1-anderson who might have some opinion here from the Drush/Drupal side.

**Comment by greg-1-anderson** [comment]
Oof. I have a lot of sympathy for Drupal folks who want to remove modules. The workflow of deploying code to live can be fraught, because you will have difficulty removing an enabled module if its code goes away.

However, it is my feeling that giving your code deployment process access to your site's database (e.g. as configuration / environment variables that a Composer script or pluign could use to uninstall modules) is not the way to solve it. I think the best deploy workflow would involve a dev -> test -> live setup. You steps should be:

 - Remove modules from the database
 - Deploy new code
 - drush updatedb
 - Run tests that include confirmation that all enabled modules appear in the codebase

If you have a deployment that requires a new feature to come online at the same time a module is being removed, then you should keep the code for that module in your code base during the new feature deploy, and uninstall it with an update hook. Removing the module should be done as a follow-on deploy. Use visual regression tests to confirm that the code removal did not alter behavior.

In summary, I would discourage folks from trying to solve this problem at the Composer level. The solution is staged deploys, and more tests.

**Comment by greg-1-anderson** [comment]
If you want to write some magic code to make a difficult process easier, how about a Drush command that uninstalls a module by writing a hook_update_N for you, and then calls `drush updatedb`? Then you'd have a convenient workflow that deployed cleanly.

e.g.
```
drush remove-on-deploy oldmodule --hook-module=mycustomsitemodule
```
You could set `--hook-module` as a command-specific option in your site-local drush.yml file, and then you wouldn't need to specify it on the commandline every time.

This is just a spur-of-the-moment idea I had. Not sure if you'd run into any issues if you tried it. Seems like it would be pretty easy to write, though.

**Comment by Seldaek** [comment]
Thanks both for the feedback. I hope this is helpful @geek-merlin. I'm going to close this here as I don't think it's something we can act on really.

**Comment by Seldaek** [comment]
@geek-merlin you should be able to listen to the `pre-operations-exec` event in a plugin and update the operation list to remove any UninstallOperation that you don't want.

**Comment by geek-merlin** [comment]
Thanks @Seldaek this helps me a lot.
So let's pioneer in contrib if this is brainfart or genuis.
